File: mobile/iOS/Sources/Store/CatalogueCache.swift
import Foundation

/// Disk cache of catalogue JSON so Home, Library and artist pages keep working
/// with the radio off. Mirrors Android `CatalogueCache`.
enum CatalogueCache {
    private static var directory: URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? FileManager.default.temporaryDirectory
        let dir = base.appendingPathComponent("catalogue", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    static func read<T: Decodable>(_ type: T.Type, key: String) -> T? {
        let url = file(for: key)
        guard let data = try? Data(contentsOf: url) else { return nil }
        return try? JSONDecoder().decode(T.self, from: data)
    }

    static func write<T: Encodable>(_ value: T, key: String) {
        guard let data = try? JSONEncoder().encode(value) else { return }
        try? data.write(to: file(for: key), options: .atomic)
    }

    static func clear() {
        try? FileManager.default.removeItem(at: directory)
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    }

    private static func file(for key: String) -> URL {
        let safe = key.replacingOccurrences(of: "[^a-zA-Z0-9._-]", with: "_", options: .regularExpression)
        return directory.appendingPathComponent("\(safe).json")
    }

    static let feed = "feed"
    static let dailyMixes = "daily-mixes"
    static let genres = "genres"
    static let liked = "liked"
    static let playlists = "playlists"
    static let followed = "followed"
    static let recent = "recent"
    static let albums = "albums"
    static let profile = "profile"
    static func artist(_ id: String) -> String { "artist-\(id)" }
    static func artistTracks(_ id: String) -> String { "artist-tracks-\(id)" }
    static func album(_ id: String) -> String { "album-\(id)" }
    static func playlist(_ id: String) -> String { "playlist-\(id)" }
}

File: mobile/android/app/src/main/java/app/serika/musicy/mobile/data/cache/CatalogueCache.kt
package app.serika.musicy.mobile.data.cache

import android.content.Context
import app.serika.musicy.mobile.data.api.ApiClient
import java.io.File
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.KSerializer

/**
 * Disk cache of catalogue JSON so Home, Library and artist/album pages keep
 * working with the radio off. Each key is a file under filesDir/catalogue.
 */
class CatalogueCache(context: Context) {

    private val directory = File(context.applicationContext.filesDir, "catalogue").apply { mkdirs() }
    private val json = ApiClient.json
    private val mutex = Mutex()

    suspend fun <T> read(key: String, serializer: KSerializer<T>): T? = withContext(Dispatchers.IO) {
        mutex.withLock {
            val file = fileFor(key)
            if (!file.exists() || file.length() == 0L) return@withLock null
            runCatching { json.decodeFromString(serializer, file.readText()) }.getOrNull()
        }
    }

    suspend fun <T> write(key: String, value: T, serializer: KSerializer<T>) = withContext(Dispatchers.IO) {
        mutex.withLock {
            runCatching {
                fileFor(key).writeText(json.encodeToString(serializer, value))
            }
        }
        Unit
    }

    suspend fun remove(key: String) = withContext(Dispatchers.IO) {
        mutex.withLock { fileFor(key).delete() }
        Unit
    }

    suspend fun clear() = withContext(Dispatchers.IO) {
        mutex.withLock { directory.listFiles()?.forEach { it.delete() } }
        Unit
    }

    fun sizeBytes(): Long = directory.walkTopDown().filter { it.isFile }.sumOf { it.length() }

    private fun fileFor(key: String): File {
        val safe = key.replace(Regex("[^a-zA-Z0-9._-]"), "_")
        return File(directory, "$safe.json")
    }

    companion object {
        const val FEED = "feed"
        const val DAILY_MIXES = "daily-mixes"
        const val GENRES = "genres"
        const val LIKED = "liked"
        const val PLAYLISTS = "playlists"
        const val FOLLOWED = "followed"
        const val RECENT = "recent"
        const val ALBUMS = "albums"
        const val ARTISTS = "artists"
        const val PROFILE = "profile"
        fun artist(id: String) = "artist-$id"
        fun artistTracks(id: String) = "artist-tracks-$id"
        fun artistAlbums(id: String) = "artist-albums-$id"
        fun album(id: String) = "album-$id"
        fun playlist(id: String) = "playlist-$id"
        fun mix(id: String) = "mix-$id"
        fun track(id: String) = "track-$id"
    }
}
